homeassistant/components/vacuum/mqtt.py

```diff
@@ -0,0 +1,251 @@
+        self._payload_locate = "locate"
+        self._payload_start_pause = "start_pause"
+        self._retain = False
+
+        self._state_template = state_template
+
+        self._state = False
+        self._status = 'Unknown'
+        self._battery_level = 0
+
+    @asyncio.coroutine
+    def async_added_to_hass(self):
+        """Subscribe MQTT events.
+
+        This method is a coroutine.
+        """
+        @callback
+        def message_received(topic, payload, qos):
+            """Handle new MQTT message."""
+            try:
+                payload = json.loads(payload)
+            except ValueError:  # e.g. json.decoder.JSONDecodeError:
+                _LOGGER.warning("JSONDecodeError in vacuum.mqtt payload: %s",
+                                payload)
+                return
+
+            battery_level = payload.get(ATTR_BATTERY_LEVEL)
+            if battery_level is not None:
+                self._battery_level = int(battery_level)
+
+            charging = payload.get("charging")
+            state = payload.get(ATTR_STATE)
+            if payload[ATTR_STATE]:
+                if state == "cleaning":
+                    self._state = True
+                    self._status = "Cleaning"
+                if state == "docked":
+                    self._state = False
+                    if charging:
+                        self._status = "Docked & Charging"
+                    else:
+                        self._status = "Docked"
+                if state == "stopped":
+                    self._state = False
+                    self._status = "Stopped"
+
+            self.hass.async_add_job(self.async_update_ha_state())
+
+        yield from mqtt.async_subscribe(
+            self.hass, self._state_topic, message_received, self._qos)
+
+    @property
+    def name(self):
+        """Return the name of the vacuum."""
+        return self._name
+
+    @property
+    def icon(self):
+        """Return the icon for the vacuum."""
+        return DEFAULT_ICON
+
+    @property
+    def should_poll(self):
+        """No polling needed for an MQTT vacuum."""
+        return False
+
+    @property
+    def is_on(self):
+        """Return true if vacuum is on."""
+        return self._state
+
+    @property
+    def status(self):
+        """Return the status of the vacuum."""
+        if self.supported_features & SUPPORT_STATUS == 0:
+            return
+
+        return self._status
+
+    @property
+    def battery_level(self):
+        """Return the status of the vacuum."""
+        if self.supported_features & SUPPORT_BATTERY == 0:
+            return
+
+        return max(0, min(100, self._battery_level))
+
+    @property
+    def device_state_attributes(self):
+        """Return device state attributes."""
+        return {}
+
+    @property
+    def supported_features(self):
+        """Flag supported features."""
+        return self._supported_features
+
+    def turn_on(self, **kwargs):
```

Comment:
Change to the async version, and annotate the method with `@asyncio.coroutine`. See eg https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/switch/mqtt.py#L165 and https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/vacuum/xiaomi.py#L231.

This applies to all methods below.

Reply:
Will do, thanks for the links!

Reply:
I noticed the superclass `VacuumDevice`, has code like this

```
    def stop(self, **kwargs):
        """Stop the vacuum cleaner."""
        raise NotImplementedError()

    def async_stop(self, **kwargs):
        """Stop the vacuum cleaner.

        This method must be run in the event loop and returns a coroutine.
        """
        return self.hass.async_add_job(partial(self.stop, **kwargs))
```

Does it make sense to update the superclass code to remove the non-async version?

Reply:
No, each platform will decide to use either the sync or the async version of the method, so one of them will be overwritten. The service handler in the base component will always use the async version.